### PR TITLE
make typeguard work on python3.9

### LINKED ISSUE #149: Make typeguard work on python 3.9
Currently 26 testcases fail on 3.9
```
platform darwin -- Python 3.9.0, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
cachedir: .tox/py39/.pytest_cache
rootdir: /Users/death0wl/typeguard, configfile: setup.cfg, testpaths: tests
plugins: typeguard-2.9.1.post8+dirty
collected 231 items

tests/test_importhook.py ................                                                                                                                                                                                            [  6%]
tests/test_typeguard.py ...........F.........F.......................FF....................................F..........F.F..F.FFFFF..F.F.F.............F.F..F.F......................F.F.F............FF....F.......................  [ 94%]
tests/test_typeguard_py36.py ............                                                                                                                                                                                            [100%]

================================================================================================================= FAILURES =================================================================================================================
________________________________________________________________________________________________________ test_check_recursive_type _________________________________________________________________________________________________________
tests/test_typeguard.py:105: in test_check_recursive_type
    check_type('foo', {'a': [1, 2, 3]}, JSONType)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:426: in check_union
    raise TypeError('type of {} must be one of ({}); got {} instead'.
E   TypeError: type of foo must be one of (str, int, float, bool, NoneType, List, Dict); got dict instead
________________________________________________________________________________________________ TestCheckArgumentTypes.test_callable_plain ________________________________________________________________________________________________
tests/test_typeguard.py:198: in test_callable_plain
    foo(callback)
tests/test_typeguard.py:193: in foo
    assert check_argument_types()
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:250: in check_callable
    if expected_type.__args__:
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
___________________________________________________________________________________________ TestCheckArgumentTypes.test_namedtuple_type_mismatch ___________________________________________________________________________________________
tests/test_typeguard.py:416: in test_namedtuple_type_mismatch
    pytest.raises(TypeError, foo, ('bob', 1)).\
E   Failed: DID NOT RAISE <class 'TypeError'>
_________________________________________________________________________________________ TestCheckArgumentTypes.test_namedtuple_wrong_field_type __________________________________________________________________________________________
tests/test_typeguard.py:426: in test_namedtuple_wrong_field_type
    pytest.raises(TypeError, foo, Employee(2, 1)).\
E   Failed: DID NOT RAISE <class 'TypeError'>
________________________________________________________________________________________________ TestCheckArgumentTypes.test_recursive_type ________________________________________________________________________________________________
tests/test_typeguard.py:715: in test_recursive_type
    foo({'a': [1, 2, 3]})
tests/test_typeguard.py:713: in foo
    assert check_argument_types()
typeguard/__init__.py:724: in check_argument_types
    raise exc from None
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:426: in check_union
    raise TypeError('type of {} must be one of ({}); got {} instead'.
E   TypeError: type of argument "arg" must be one of (str, int, float, bool, NoneType, List, Dict); got dict instead
______________________________________________________________________________________________ TestTypeChecked.test_callable[unparametrized] _______________________________________________________________________________________________
tests/test_typeguard.py:801: in test_callable
    foo(some_callable)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:250: in check_callable
    if expected_type.__args__:
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
________________________________________________________________________________________________ TestTypeChecked.test_list[unparametrized] _________________________________________________________________________________________________
tests/test_typeguard.py:813: in test_list
    foo([1, 2])
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:335: in check_list
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
________________________________________________________________________________________________ TestTypeChecked.test_dict[unparametrized] _________________________________________________________________________________________________
tests/test_typeguard.py:825: in test_dict
    foo({'x': 2})
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:300: in check_dict
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_______________________________________________________________________________________ TestTypeChecked.test_unparametrized_types_mixed_values[dict] _______________________________________________________________________________________
tests/test_typeguard.py:840: in test_unparametrized_types_mixed_values
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:300: in check_dict
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_______________________________________________________________________________________ TestTypeChecked.test_unparametrized_types_mixed_values[list] _______________________________________________________________________________________
tests/test_typeguard.py:840: in test_unparametrized_types_mixed_values
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:335: in check_list
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_____________________________________________________________________________________ TestTypeChecked.test_unparametrized_types_mixed_values[sequence] _____________________________________________________________________________________
tests/test_typeguard.py:840: in test_unparametrized_types_mixed_values
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:347: in check_sequence
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_______________________________________________________________________________________ TestTypeChecked.test_unparametrized_types_mixed_values[set] ________________________________________________________________________________________
tests/test_typeguard.py:840: in test_unparametrized_types_mixed_values
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
___________________________________________________________________________________ TestTypeChecked.test_unparametrized_types_mixed_values[abstractset] ____________________________________________________________________________________
tests/test_typeguard.py:840: in test_unparametrized_types_mixed_values
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
___________________________________________________________________________________________ TestTypeChecked.test_sequence[tuple-unparametrized] ____________________________________________________________________________________________
tests/test_typeguard.py:853: in test_sequence
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:347: in check_sequence
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
____________________________________________________________________________________________ TestTypeChecked.test_sequence[list-unparametrized] ____________________________________________________________________________________________
tests/test_typeguard.py:853: in test_sequence
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:347: in check_sequence
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
____________________________________________________________________________________________ TestTypeChecked.test_sequence[str-unparametrized] _____________________________________________________________________________________________
tests/test_typeguard.py:853: in test_sequence
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:347: in check_sequence
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
________________________________________________________________________________________________ TestTypeChecked.test_set[value0-abstract] _________________________________________________________________________________________________
tests/test_typeguard.py:894: in test_set
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_____________________________________________________________________________________________ TestTypeChecked.test_set[value0-unparametrized] ______________________________________________________________________________________________
tests/test_typeguard.py:894: in test_set
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
________________________________________________________________________________________________ TestTypeChecked.test_set[value1-abstract] _________________________________________________________________________________________________
tests/test_typeguard.py:894: in test_set
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
_____________________________________________________________________________________________ TestTypeChecked.test_set[value1-unparametrized] ______________________________________________________________________________________________
tests/test_typeguard.py:894: in test_set
    foo(value)
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:360: in check_set
    if expected_type.__args__ not in (None, expected_type.__parameters__):
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
______________________________________________________________________________________________ TestTypeChecked.test_generator[bare_generator] ______________________________________________________________________________________________
tests/test_typeguard.py:1015: in test_generator
    gen = genfunc()
typeguard/__init__.py:873: in wrapper
    return TypeCheckedGenerator(retval, memo)
typeguard/__init__.py:731: in __init__
    rtype_args = memo.type_hints['return'].__args__
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
______________________________________________________________________________________________ TestTypeChecked.test_generator[bare_iterable] _______________________________________________________________________________________________
tests/test_typeguard.py:1015: in test_generator
    gen = genfunc()
typeguard/__init__.py:873: in wrapper
    return TypeCheckedGenerator(retval, memo)
typeguard/__init__.py:731: in __init__
    rtype_args = memo.type_hints['return'].__args__
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
______________________________________________________________________________________________ TestTypeChecked.test_generator[bare_iterator] _______________________________________________________________________________________________
tests/test_typeguard.py:1015: in test_generator
    gen = genfunc()
typeguard/__init__.py:873: in wrapper
    return TypeCheckedGenerator(retval, memo)
typeguard/__init__.py:731: in __init__
    rtype_args = memo.type_hints['return'].__args__
../.pyenv/versions/3.9.0/lib/python3.9/typing.py:648: in __getattr__
    raise AttributeError(attr)
E   AttributeError: __args__
______________________________________________________________________________________________ TestTypeChecked.test_typevar_forwardref[bound] ______________________________________________________________________________________________
tests/test_typeguard.py:1165: in test_typevar_forwardref
    func(Parent())
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:724: in check_argument_types
    raise exc from None
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:649: in check_type
    check_typevar(argname, value, expected_type, memo)
typeguard/__init__.py:452: in check_typevar
    bound_type = resolve_forwardref(memo.typevars.get(typevar, typevar.__bound__), memo)
typeguard/__init__.py:176: in resolve_forwardref
    return evaluate_forwardref(maybe_ref, memo.globals, memo.locals)
E   TypeError: _evaluate() missing 1 required positional argument: 'recursive_guard'
___________________________________________________________________________________________ TestTypeChecked.test_typevar_forwardref[constrained] ___________________________________________________________________________________________
tests/test_typeguard.py:1165: in test_typevar_forwardref
    func(Parent())
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:724: in check_argument_types
    raise exc from None
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:649: in check_type
    check_typevar(argname, value, expected_type, memo)
typeguard/__init__.py:459: in check_typevar
    constraints = [resolve_forwardref(c, memo) for c in typevar.__constraints__]
typeguard/__init__.py:459: in <listcomp>
    constraints = [resolve_forwardref(c, memo) for c in typevar.__constraints__]
typeguard/__init__.py:176: in resolve_forwardref
    return evaluate_forwardref(maybe_ref, memo.globals, memo.locals)
E   TypeError: _evaluate() missing 1 required positional argument: 'recursive_guard'
___________________________________________________________________________________________________ TestTypeChecked.test_recursive_type ____________________________________________________________________________________________________
tests/test_typeguard.py:1202: in test_recursive_type
    foo({'a': [1, 2, 3]})
typeguard/__init__.py:863: in wrapper
    check_argument_types(memo)
typeguard/__init__.py:724: in check_argument_types
    raise exc from None
typeguard/__init__.py:722: in check_argument_types
    check_type(description, value, expected_type, memo)
typeguard/__init__.py:617: in check_type
    checker_func(argname, value, expected_type, memo)
typeguard/__init__.py:426: in check_union
    raise TypeError('type of {} must be one of ({}); got {} instead'.
E   TypeError: type of argument "arg" must be one of (str, int, float, bool, NoneType, List, Dict); got dict instead
====================================================================================================== 26 failed, 205 passed in 5.08s ======================================================================================================
```

### PR DESCRIPTION
**Pull request by deathowl**
Add python 3.9 to tox.ini
make tests pass for 3.9 as well as other supported python versions
fix #149

### CODE REVIEW

**Comment by deathowl** [comment]
sorry for the python3.9-dev in the travis.yml can be replaced with non-dev 3.9 once travis makes it available

File: typeguard/__init__.py [hunk 1]
```diff
@@ -436,7 +454,8 @@ def check_class(argname: str, value, expected_type, memo: _TypeCheckMemo) -> Non
     if expected_type is Type:
         return
 
-    expected_class = expected_type.__args__[0] if expected_type.__args__ else None
+    expected_class = expected_type.__args__[0] if hasattr(
```
Comment (agronholm) [inline typeguard/__init__.py:457]: If a tertiary expression has to be split to multiple lines then it should be converted to a real "if" block.

File: typeguard/__init__.py [hunk 2]
```diff
@@ -728,10 +747,11 @@ def check_argument_types(memo: Optional[_CallMemo] = None) -> bool:
 
 class TypeCheckedGenerator:
     def __init__(self, wrapped: Generator, memo: _CallMemo):
-        rtype_args = memo.type_hints['return'].__args__
+        rtype_args = memo.type_hints['return'].__args__ if hasattr(
```
Comment (agronholm) [inline typeguard/__init__.py:750]: Same here.

File: typeguard/__init__.py [hunk 3]
```diff
@@ -297,7 +301,8 @@ def check_dict(argname: str, value, expected_type, memo: _TypeCheckMemo) -> None
                         format(argname, qualified_name(value)))
 
     if expected_type is not dict:
-        if expected_type.__args__ not in (None, expected_type.__parameters__):
+        if hasattr(expected_type, "__args__") and expected_type.__args__ not in \
```
Comment (agronholm) [inline typeguard/__init__.py:304]: Parentheses are preferred over splitting lines with a backslash.

File: typeguard/__init__.py [hunk 4]
```diff
@@ -247,7 +251,7 @@ def check_callable(argname: str, value, expected_type, memo: _TypeCheckMemo) ->
     if not callable(value):
         raise TypeError('{} must be a callable'.format(argname))
 
-    if expected_type.__args__:
+    if hasattr(expected_type, "__args__") and expected_type.__args__:
```
Comment (agronholm) [inline typeguard/__init__.py:254]: `if getattr(expected_type, '__args__', None):`

**Review by agronholm — CHANGES_REQUESTED** [review]
Mostly code style changes requested. Seems good otherwise.

File: tests/test_typeguard.py [hunk 5]
```diff
@@ -10,7 +10,17 @@
     Container, Generic, BinaryIO, TextIO, Generator, Iterator, AbstractSet, AnyStr, Type)
 
 import pytest
-from typing_extensions import NoReturn, Protocol, Literal, TypedDict, runtime_checkable
```
[the remark this answers is not in the document]
Comment (deathowl) [reply tests/test_typeguard.py:13]:
```
Python 3.9.0 (default, Oct  6 2020, 10:37:47)
[Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing_extensions
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'typing_extensions'
```

[the remark this answers is not in the document]
Comment (deathowl) [reply tests/test_typeguard.py:13]: ok, removed and still works fine, i was just overly cautious

**Comment by deathowl** [comment]
@agronholm  addressed your requests.

**Review by agronholm — APPROVED** [review]

**Comment by agronholm** [comment]
I screwed up the build. I'll fix that before I merge.

**Comment by agronholm** [comment]
Thanks!

**Comment by deathowl** [comment]
btw what is your opinion on using black to automate code formatting in the future?

**Comment by agronholm** [comment]
Over my rotting corpse.

**Comment by leycec** [comment]
⮤ This is the only valid response to that question.

**Comment by shadchin** [comment]
Can you upload to PyPI release with support Python 3.9?

**Comment by agronholm** [comment]
I tried...but it seems some changes on Travis have broken my workflow. It ended uploading typeguard-0.0.0 to PyPI :(

**Comment by agronholm** [comment]
Scratch that, it was the continued modernization of setup.py.

**Comment by agronholm** [comment]
Done, it's now on PyPI.

**Pull request merged**